flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1220,8 +1224,15 @@ private void completePendingCheckpoint(PendingCheckpoint pendingCheckpoint)
             Preconditions.checkState(pendingCheckpoint.isDisposed() && completedCheckpoint != null);
 
             try {
-                completedCheckpointStore.addCheckpoint(
-                        completedCheckpoint, checkpointsCleaner, this::scheduleTriggerRequest);
+                CompletedCheckpoint lastSubsumed =
+                        completedCheckpointStore.addCheckpointAndSubsumeOldestOne(
+                                completedCheckpoint,
+                                checkpointsCleaner,
+                                this::scheduleTriggerRequest);
+                if (lastSubsumed != null && lastSubsumed.discardOnSubsume()) {
```

Comment:
I think this needs more discussion here. After we store savepoint in checkpoint store, the actual meaning of `CompletedCheckpointStore#addCheckpoint` has changed: we might subsume a savepoint and a checkpoint in the same turn when adding a new checkpoint. My understanding is that we would not need to notify tasks of subsumed savepoint as it should not be owned on task side.

Reply:
I think we shouldn't make such a distinction.
If TM needs to distinguish for some particular task (state management), it can be done on TM side.
One practical reason to notify TMs is that subsuming a savepoint allows them subsume older checkpoints.
A more general reason that the community I think is moving towards miminizing the difference in handling of savepoints and checkpoints.

Reply:
If we also send notification of subsumed savepoint, there would be a problem that we might subsume a savepoint and a checkpoint in the same turn when adding a new checkpoint. For this case, which checkpoint id should we notify?

I see your PR of https://github.com/apache/flink/pull/16575 also consider the savepoint case on TM side, was that the reason you prefer to send notification of savepoints?

Reply:
> which checkpoint id should we notify?

I think it should always be the latest (regardless whether it's savepoint or checkpoint).

> I see your PR of #16575 also consider the savepoint case on TM side, was that the reason you prefer to send notification of savepoints?

No, I mentioned the reasons above (the main one is generality). 

Regarding the PR (#16575), TM needs to distinguish when discarding the state anyways. This is because it may have non-confirmed savepoints (that might be aborted); at the same, a newer checkpoint might be subsumed; which would mean that those older savepoints are subsumed.